topic_dynamics/parsing.py

```diff
@@ -0,0 +1,376 @@
+    :param output_dir: path to the output directory.
+    :param name: name of the processed dataset.
+    :param dates: a list of dates used for slicing.
+    :return: None.
+    """
+    print("Calculating the diffs between versions and transforming the token lists.")
+    diff_indices = {}
+    count_index_diff = 0
+    with open(os.path.abspath(os.path.join(output_dir, name + "_diffs_tokens.txt")), "w+") as fout:
+        for date in tqdm(range(2, len(dates) + 1)):
+            starting_index_diff = count_index_diff + 1
+            # Save the tokens of the "previous" slice into memory
+            previous_version = {}
+            with open(os.path.abspath(os.path.join(slices_tokens_dir, str(date - 1) + ".txt"))) as fin:
+                for line in fin:
+                    previous_version[line.rstrip().split(";")[1]] = line.rstrip().split(";")[2]
+            current_version = []
+            with open(os.path.abspath(os.path.join(slices_tokens_dir, str(date) + ".txt"))) as fin:
+                for line in fin:
+                    # Iterate over files in the "current" version
+                    current_version.append(line.rstrip().split(";")[1])
+                    address = line.rstrip().split(";")[1]
+                    tokens = {}
+                    for token in line.rstrip().split(";")[2].split(","):
+                        tokens[token.split(":")[0]] = int(token.split(":")[1])
+                    tokens = Counter(tokens)
+                    old_address = address.replace(dates[date - 1].strftime("%Y-%m-%d"),
+                                                  dates[date - 2].strftime("%Y-%m-%d"), 1)
+                    # Check if the file with this name existed in the previous version
+                    if old_address in previous_version.keys():
+                        old_tokens = {}
+                        for token in previous_version[old_address].split(","):
+                            old_tokens[token.split(":")[0]] = int(token.split(":")[1])
+                        old_tokens = Counter(old_tokens)
+                        # Calcualate which tokens have been added and removed between versions
+                        created_tokens = sorted((tokens - old_tokens).items(), key=itemgetter(1), reverse=True)
+                        deleted_tokens = sorted((old_tokens - tokens).items(), key=itemgetter(1), reverse=True)
+                        new_tokens = []
+                        if len(created_tokens) != 0:
+                            for token in created_tokens:
+                                new_token = "+" + token[0]
+                                new_tokens.append([new_token, token[1]])
+                        if len(deleted_tokens) != 0:
+                            for token in deleted_tokens:
+                                new_token = "-" + token[0]
+                                new_tokens.append([new_token, token[1]])
```

Comment:
```suggestion
                        token_cnt_updates = (tokens - old_tokens).items()
                        new_tokens = []
                        if token_updates:
                            for token_cnt in created_tokens:
                               sign = "+" if token_cnt[1] > 0 else "-"
                               new_tokens.append([sign + token_cnt[0], token_cnt[1]])
```

Reply:
Substracting two Counter objects doesn't leave negative options or zeroes. That's why I used it. It can't store negative values. There is a way to substact it to keep zeroes and negative values, but it adds a hell of a lot of other problems of iterating - becomes really hard. These two differences actually seem to be a more elegant solution, of those that I have found.